clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internal/expiring/ExpiringCredentialRefreshingLogin.java

```diff
@@ -0,0 +1,382 @@
+                 */
+                long nowMs = currentMs();
+                Long nextRefreshMs = getRefreshMs(nowMs);
+                if (nextRefreshMs == null) {
+                    loginContextFactory.refresherThreadDone();
+                    return;
+                }
+                log.info("[Principal={}]: Expiring credential re-login sleeping until: {}", getPrincipalLogText(),
+                        new Date(nextRefreshMs));
+                try {
+                    time.sleep(nextRefreshMs - nowMs);
+                    if (Thread.currentThread().isInterrupted())
+                        throw new InterruptedException();
+                } catch (InterruptedException ie) {
+                    log.warn("[Principal={}]: Expiring credential re-login thread has been interrupted and will exit.",
+                            getPrincipalLogText());
+                    loginContextFactory.refresherThreadDone();
+                    return;
+                }
+                try {
+                    while (true) {
+                        try {
+                            reLogin();
+                            break;
+                        } catch (LoginException loginException) {
+                            long secondsToSleep = 10L;
+                            log.warn(String.format(
+                                    "[Principal=%s]: LoginException during login retry; will sleep %d seconds before trying again.",
+                                    getPrincipalLogText(), secondsToSleep), loginException);
+                            // sleep and swallow original login exception unless interrupted
+                            try {
+                                time.sleep(secondsToSleep * 1000L);
+                                if (Thread.currentThread().isInterrupted())
+                                    throw new InterruptedException();
+                            } catch (InterruptedException e) {
+                                log.error("[Principal={}]: Interrupted during login retry after LoginException: {}",
+                                        getPrincipalLogText(), String.valueOf(loginException.getMessage()));
+                                throw loginException;
+                            }
+                        }
+                    }
+                } catch (LoginException le) {
+                    log.error(String.format(
+                            "[Principal=%s]: Interrupted while trying to perform a subsequent expiring credential re-login after one or more initial re-login failures: re-login thread exiting now.",
+                            getPrincipalLogText()), le);
+                    loginContextFactory.refresherThreadDone();
+                    return;
+                }
+            }
+        }
+
+        private void reLogin() throws LoginException {
+            // Only perform one refresh at a time
+            synchronized (getClass()) {
```

Comment:
I've adjusted this class since this review.  The optional class to synchronize on is now provided in the constructor.  This allows the specific mechanism to determine if the relogins should be serialized or not.  Is the new version acceptable?

Reply:
Is there a case when we don't need synchronization? Passing in the class from the constructor is fine, but I think we should just synchronize all the time.